torch/lib/c10d/logger.cpp

```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
   LogPyTorchDDPUsage(*ddp_logging_data_);
 }
 
+void Logger::calculate_avg_cpu_time(
+    int64_t& avg_time,
+    int64_t cpu_start_time,
+    int64_t cpu_end_time) {
+  // If cpu_end_time is not recorded in this iteration,
+  // avg_time will return invalid value.
+  // For some cases like DDP runs on non-sync mode, backward compute
+  // end time can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  TORCH_CHECK(reducer_->num_iterations_ > 0);
+  if (cpu_end_time < cpu_start_time) {
+    return;
+  }
+  long num_iters = reducer_->num_iterations_;
+  avg_time = ((cpu_end_time - cpu_start_time) + avg_time * (num_iters - 1)) /
+      num_iters;
+}
+
+#ifdef USE_CUDA
+void Logger::calculate_avg_gpu_time(
+    int64_t& avg_time,
+    at::cuda::CUDAEvent& gpu_start,
+    at::cuda::CUDAEvent& gpu_end) {
+  TORCH_CHECK(reducer_->num_iterations_ > 0);
+  long num_iters = reducer_->num_iterations_;
+  float milliseconds = gpu_start.elapsed_time(gpu_end);
+  // If gpu_end is not recorded in this iteration,
+  // milliseconds will have invalid value.
+  // For some cases like DDP runs on non-sync mode,
+  // gpu_end can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  if (milliseconds < 0) {
+    return;
+  }
+  avg_time = (int(milliseconds * kMilliSecondToNanosSecond) +
+              avg_time * (num_iters - 1)) /
+      num_iters;
+}
+#endif
+
+void Logger::set_runtime_stats() {
+  // Sync with reducer's data
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
+  // set runtime stats after 1st iteration is complete.
+  if (reducer_->num_iterations_ == 0) {
+    return;
+  }
+  // Set ith iteration when the runtime stats are set.
+  ddp_logging_data_->iteration = reducer_->num_iterations_;
+  // If unused_parameters_ is not empty, calculate its sizes.
+  // unused_parameters_ is calculated in forward call of
+  // each iteration.
+  for (const auto& unused_index : reducer_->unused_parameters_) {
+    const auto& v = reducer_->replicas_[unused_index.replica_index]
+                                       [unused_index.variable_index];
+    ddp_logging_data_->unused_parameter_size += v.numel() * v.element_size();
+  }
+  // rebuilt_bucket_sizes will not change once buckets are rebuilt,
+  // so it only needs to set once during whole training loop.
+  if (ddp_logging_data_->has_rebuilt_buckets != reducer_->has_rebuilt_bucket_) {
+    ddp_logging_data_->has_rebuilt_buckets = reducer_->has_rebuilt_bucket_;
+    ddp_logging_data_->rebuilt_bucket_sizes = get_bucket_sizes();
+  }
+
+  if (reducer_->replicas_[0][0].is_cuda()) {
+#ifdef USE_CUDA
+    // Cuda time stats are only collected for single process single
+    // device and single device module.
+    if (reducer_->replicas_.size() > 1 || reducer_->is_multi_device_module_) {
```

Comment:
I'm assuming that multi device module is what was causing the cuda deadlock issues you mentioned? Do we know if we want to eventually support this, or is there some blocker? Could we file issues as needed?

Reply:
cuda event, record and synchronize needs to be on the same device. currently we created events on a fixed device only once during whole training loop (because event creation is expensive, we should avoid creating it every iteration), single process multiple device or multiple device module will result in "invalid resource handle" cuda error when devices are not consistent with creation device.

Reply:
(don't need to add this in this PR, just for your reference) I created an example of using `CUDAEvent` to measure multi-device module here: https://gist.github.com/mrshenli/8c39c35612218e0d6d772910bca2f737#file-event_demo-py-L57-L81

The high-level idea is to first build dependencies between streams and then create the `CUDAEvent`.

Reply:
thanks for the example, will follow up with you offline. we can add multi-device module support in the future if possible